lib/web/kube.go

```diff
@@ -0,0 +1,344 @@
+	if err != nil {
+		return trace.Wrap(err, "failed getting user private key from the session")
+	}
+	userKey := &client.Key{
+		PrivateKey: pk,
+		Cert:       p.sctx.cfg.Session.GetPub(),
+		TLSCert:    p.sctx.cfg.Session.GetTLSCert(),
+	}
+
+	hostCA, err := p.localAccessPoint.GetCertAuthority(ctx, types.CertAuthID{
+		Type:       types.HostCA,
+		DomainName: p.teleportCluster,
+	}, false)
+	if err != nil {
+		return trace.Wrap(err, "failed getting host CA")
+	}
+
+	stream := NewTerminalStream(ctx, TerminalStreamConfig{WS: p.ws, Logger: p.log})
+
+	certsReq := clientproto.UserCertsRequest{
+		PublicKey:         userKey.MarshalSSHPublicKey(),
+		Username:          p.sctx.GetUser(),
+		Expires:           p.sctx.cfg.Session.GetExpiryTime(),
+		Format:            constants.CertificateFormatStandard,
+		RouteToCluster:    p.teleportCluster,
+		KubernetesCluster: p.req.KubeCluster,
+		Usage:             clientproto.UserCertsRequest_Kubernetes,
+	}
+
+	_, certs, err := client.PerformMFACeremony(ctx, client.PerformMFACeremonyParams{
+		CurrentAuthClient: p.userClient,
+		RootAuthClient:    p.sctx.cfg.RootClient,
+		MFAPrompt: mfa.PromptFunc(func(ctx context.Context, chal *clientproto.MFAAuthenticateChallenge) (*clientproto.MFAAuthenticateResponse, error) {
+			assertion, err := promptMFAChallenge(stream.WSStream, protobufMFACodec{}).Run(ctx, chal)
+			return assertion, trace.Wrap(err)
+		}),
+		MFAAgainstRoot: p.sctx.cfg.RootClusterName == p.teleportCluster,
+		MFARequiredReq: &clientproto.IsMFARequiredRequest{
+			Target: &clientproto.IsMFARequiredRequest_KubernetesCluster{KubernetesCluster: p.req.KubeCluster},
+		},
+		ChallengeExtensions: mfav1.ChallengeExtensions{
+			Scope: mfav1.ChallengeScope_CHALLENGE_SCOPE_USER_SESSION,
+		},
+		CertsReq: &certsReq,
+		Key:      userKey,
+	})
+	if err != nil && !errors.Is(err, services.ErrSessionMFANotRequired) {
+		return trace.Wrap(err, "failed performing mfa ceremony")
+	}
+
+	if certs == nil {
+		certs, err = p.sctx.cfg.RootClient.GenerateUserCerts(ctx, certsReq)
+		if err != nil {
+			return trace.Wrap(err, "failed issuing user certs")
+		}
+	}
+
+	rsaKey, err := userKey.PrivateKey.RSAPrivateKeyPEM()
+	if err != nil {
+		return trace.Wrap(err, "failed getting rsa private key")
+	}
+
+	restConfig, err := createKubeRestConfig(p.configServerAddr, p.configTLSServerName, hostCA, certs.TLS, rsaKey)
+	if err != nil {
+		return trace.Wrap(err, "failed creating Kubernetes rest config")
+	}
+
+	kubeClient, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return trace.Wrap(err, "failed creating Kubernetes client")
+	}
+
+	kubeReq := kubeClient.CoreV1().RESTClient().Post().Resource("pods").Name(p.req.Pod).
+		Namespace(p.req.Namespace).SubResource("exec")
+	option := &v1.PodExecOptions{
+		Container: p.req.Container,
+		Command:   strings.Split(p.req.Command, " "),
+		TTY:       p.req.IsInteractive,
+		Stdin:     p.req.IsInteractive,
+		Stdout:    true,
+		Stderr:    true,
```

Comment:
if using TTY, you can only use stdout otherwise upgrade will hang

Reply:
Changed to be unset if it's interactive: https://github.com/gravitational/teleport/pull/41144/commits/db1f28660cf2ea632595823f4f3ffc9671ff6d9f